File: clients/java/src/main/java/com/docmgmt/client/ContentAndSearchDemo.java
package com.docmgmt.client;

import java.io.IOException;
import java.util.*;

/**
 * Document Management System - Content, Renditions, and Search Demo
 * 
 * This demo shows:
 * 1. Creating documents with indexable content
 * 2. Uploading PDF files
 * 3. Using transformation features (via UI)
 * 4. Searching documents with full-text and fielded search
 * 5. Rebuilding the search index
 */
public class ContentAndSearchDemo {
    
    private static final String BASE_URL = "http://localhost:8082/docmgmt/api";
    
    public static void main(String[] args) {
        System.out.println("=== Document Management: Content, Renditions & Search Demo ===\n");
        
        DocumentManagementClient client = new DocumentManagementClient(BASE_URL);
        
        try {
            // ========================================
            // PART 1: Create Documents with Content
            // ========================================
            System.out.println("📄 PART 1: Creating Documents with Indexable Content\n");
            
            // Create a document with text content (indexable)
            System.out.println("1. Creating document with text content...");
            Map<String, Object> doc1 = client.createDocument(
                "Spring Framework Tutorial",
                "ARTICLE",
                "A comprehensive guide to Spring Boot development",
                null,  // author (deprecated)
                "spring boot java framework web",
                Arrays.asList("java", "spring", "tutorial", "backend")
            );
            Long doc1Id = ((Number) doc1.get("id")).longValue();
            System.out.println("   ✓ Created document ID: " + doc1Id);
            
            // Upload indexable text content
            String textContent = """
                Spring Framework Tutorial
                ========================
                
                Spring Boot is a powerful framework for building enterprise Java applications.
                It provides features like:
                - Dependency Injection
                - Auto-configuration
                - Embedded servers
                - Spring Data for database access
                - Spring Security for authentication
                
                This makes it ideal for microservices and web applications.
                """.trim();
            
            Map<String, Object> content1 = client.uploadContentBytes(
                "spring_tutorial.txt",
                textContent.getBytes(),
                doc1Id,
                true,  // store in database
                null,  // file store ID
                "text/plain"
            );
            System.out.println("   ✓ Uploaded text content ID: " + content1.get("id"));
            
            // Create another document
            System.out.println("\n2. Creating document with PDF content...");
            Map<String, Object> doc2 = client.createDocument(
                "Python Best Practices",
                "MANUAL",
                "Python programming best practices and design patterns",
                null,
                "python programming design patterns",
                Arrays.asList("python", "best-practices", "programming")
            );
            Long doc2Id = ((Number) doc2.get("id")).longValue();
            System.out.println("   ✓ Created document ID: " + doc2Id);
            
            // Upload PDF content (simulated)
            byte[] pdfBytes = createSamplePdfBytes();
            Map<String, Object> content2 = client.uploadContentBytes(
                "python_guide.pdf",
                pdfBytes,
                doc2Id,
                true,
                null,
                "application/pdf"
            );
            System.out.println("   ✓ Uploaded PDF content ID: " + content2.get("id"));
            
            // Create a third document for search testing
            System.out.println("\n3. Creating additional document...");
            Map<String, Object> doc3 = client.createDocument(
                "Database Design Principles",
                "MANUAL",
                "SQL and NoSQL database design fundamentals",
                null,
                "database sql nosql design optimization",
                Arrays.asList("database", "sql", "design")
            );
            Long doc3Id = ((Number) doc3.get("id")).longValue();
            
            String content3Text = """
                Database Design Principles
                =========================
                
                Effective database design is crucial for application performance.
                
                Key principles:
                1. Normalization - Organize data to reduce redundancy
                2. Indexing - Speed up query performance
                3. Relationships - Define clear entity relationships
                4. Scalability - Plan for growth
                
                Both SQL and NoSQL databases have their place depending on requirements.
                """.trim();
            
            client.uploadContentBytes(
                "database_guide.txt",
                content3Text.getBytes(),
                doc3Id,
                true,
                null,
                "text/plain"
            );
            System.out.println("   ✓ Created document ID: " + doc3Id + " with text content");
            
            // ========================================
            // PART 2: Transformations & Renditions
            // ========================================
            System.out.println("\n\n📝 PART 2: Content Transformations & Renditions\n");
            
            System.out.println("NOTE: Content transformation (PDF → text) is currently done via UI.");
            System.out.println("Steps to transform PDF content:");
            System.out.println("  1. Go to Documents view in the UI");
            System.out.println("  2. Select your document");
            System.out.println("  3. In the content panel, find your PDF");
            System.out.println("  4. Click the ✨ (magic wand) Transform button");
            System.out.println("  5. The system will:");
            System.out.println("     - Extract text from the PDF using Apache PDFBox");
            System.out.println("     - Create a secondary rendition (text/plain)");
            System.out.println("     - Mark it as indexable");
            System.out.println("     - Add it to the document\n");
            
            System.out.println("After transformation, you'll see:");
            List<Map<String, Object>> contents = client.getContentBySysObject(doc2Id);
            System.out.println("   Document " + doc2Id + " currently has " + contents.size() + " content item(s)");
            for (Map<String, Object> c : contents) {
                System.out.println("   - " + c.get("name") + " (" + c.getOrDefault("contentType", "unknown") + ")");
            }
            
            System.out.println("\n   After transformation in UI, you'll have:");
            System.out.println("   - python_guide.pdf (application/pdf) [Primary, Not Indexable]");
            System.out.println("   - python_guide.pdf.plain (text/plain) [Secondary, Indexable]");
            
            // ========================================
            // PART 3: Search Index Management
            // ========================================
            System.out.println("\n\n🔍 PART 3: Search Index Management\n");
            
            System.out.println("1. Rebuilding search index...");
            String result = client.rebuildSearchIndex();
            System.out.println("   ✓ " + result);
            
            System.out.println("\n2. Checking index statistics...");
            Map<String, Object> stats = client.getSearchStats();
            System.out.println("   Documents indexed: " + stats.getOrDefault("documentCount", 0));
            System.out.println("   Max doc ID: " + stats.getOrDefault("maxDoc", 0));
            System.out.println("   Deleted docs: " + stats.getOrDefault("deletedDocs", 0));
            
            // ========================================
            // PART 4: Search Operations
            // ========================================
            System.out.println("\n\n🔎 PART 4: Searching Documents\n");
            
            // Simple search
            System.out.println("1. Simple search for 'spring'...");
            List<Map<String, Object>> results = client.search("spring", 10);
            System.out.println("   Found " + results.size() + " result(s):");
            for (Map<String, Object> r : results) {
                System.out.printf("   - Doc %d: %s (score: %.2f)%n",
                    ((Number) r.get("documentId")).longValue(),
                    r.getOrDefault("name", "N/A"),
                    ((Number) r.get("score")).floatValue());
            }
            
            // Search in description
            System.out.println("\n2. Search for 'framework'...");
            results = client.search("framework", 10);
            System.out.println("   Found " + results.size() + " result(s):");
            for (Map<String, Object> r : results) {
                System.out.printf("   - Doc %d: %s%n",
                    ((Number) r.get("documentId")).longValue(),
                    r.getOrDefault("name", "N/A"));
            }
            
            // Search in content
            System.out.println("\n3. Search for 'microservices' (in content)...");
            results = client.search("microservices", 10);
            System.out.println("   Found " + results.size() + " result(s):");
            for (Map<String, Object> r : results) {
                System.out.printf("   - Doc %d: %s%n",
                    ((Number) r.get("documentId")).longValue(),
                    r.getOrDefault("name", "N/A"));
            }
            
            // Fielded search - tags only
            System.out.println("\n4. Field-specific search (tags='python')...");
            Map<String, String> fieldQueries = new HashMap<>();
            fieldQueries.put("tags", "python");
            results = client.searchFields(fieldQueries, "AND", 10);
            System.out.println("   Found " + results.size() + " result(s):");
            for (Map<String, Object> r : results) {
                System.out.printf("   - Doc %d: %s%n",
                    ((Number) r.get("documentId")).longValue(),
                    r.getOrDefault("name", "N/A"));
                if (r.containsKey("tags")) {
                    System.out.println("     Tags: " + r.get("tags"));
                }
            }
            
            // Multiple field search with AND
            System.out.println("\n5. Multi-field search (name='database' AND keywords='sql')...");
            fieldQueries = new HashMap<>();
            fieldQueries.put("name", "database");
            fieldQueries.put("keywords", "sql");
            results = client.searchFields(fieldQueries, "AND", 10);
            System.out.println("   Found " + results.size() + " result(s):");
            for (Map<String, Object> r : results) {
                System.out.printf("   - Doc %d: %s%n",
                    ((Number) r.get("documentId")).longValue(),
                    r.getOrDefault("name", "N/A"));
            }
            
            // Phrase search
            System.out.println("\n6. Phrase search for 'best practices'...");
            results = client.search("\"best practices\"", 10);
            System.out.println("   Found " + results.size() + " result(s):");
            for (Map<String, Object> r : results) {
                System.out.printf("   - Doc %d: %s%n",
                    ((Number) r.get("documentId")).longValue(),
                    r.getOrDefault("name", "N/A"));
            }
            
            // Wildcard search
            System.out.println("\n7. Wildcard search for 'databa*'...");
            results = client.search("databa*", 10);
            System.out.println("   Found " + results.size() + " result(s):");
            for (Map<String, Object> r : results) {
                System.out.printf("   - Doc %d: %s%n",
                    ((Number) r.get("documentId")).longValue(),
                    r.getOrDefault("name", "N/A"));
            }
            
            // ========================================
            // PART 5: Working with Renditions
            // ========================================
            System.out.println("\n\n📦 PART 5: Inspecting Content Renditions\n");
            
            System.out.println("1. Content items for document " + doc1Id + " (text document):");
            List<Map<String, Object>> doc1Contents = client.getContentBySysObject(doc1Id);
            for (Map<String, Object> c : doc1Contents) {
                System.out.println("   - " + c.get("name"));
                System.out.println("     Type: " + c.getOrDefault("contentType", "unknown"));
                System.out.println("     Storage: " + c.getOrDefault("storageType", "unknown"));
            }
            
            System.out.println("\n2. Content items for document " + doc2Id + " (PDF document):");
            List<Map<String, Object>> doc2Contents = client.getContentBySysObject(doc2Id);
            for (Map<String, Object> c : doc2Contents) {
                System.out.println("   - " + c.get("name"));
                System.out.println("     Type: " + c.getOrDefault("contentType", "unknown"));
            }
            
            // ========================================
            // Summary
            // ========================================
            System.out.println("\n\n" + "=".repeat(60));
            System.out.println("📊 DEMO SUMMARY");
            System.out.println("=".repeat(60));
            System.out.println("✓ Created 3 documents");
            System.out.println("✓ Uploaded 3 content items");
            System.out.println("✓ Demonstrated search across multiple fields");
            System.out.println("✓ Showed fielded search with AND/OR operators");
            System.out.println("✓ Used phrase and wildcard search");
            System.out.println("\n💡 KEY FEATURES:");
            System.out.println("   • All document fields are indexed (name, description, keywords, tags)");
            System.out.println("   • Text content is automatically indexed");
            System.out.println("   • PDF → text transformation via UI (creates indexable renditions)");
            System.out.println("   • Primary/secondary rendition support");
            System.out.println("   • Full Lucene query syntax support");
            System.out.println("   • Field-specific and combined search");
            System.out.println("\n📝 NEXT STEPS:");
            System.out.println("   1. Transform PDF content via UI to create text renditions");
            System.out.println("   2. Rebuild index after transformations");
            System.out.println("   3. Search will include PDF text content");
            System.out.println("   4. Try advanced Lucene queries: 'spring AND java', 'name:Tutorial', etc.");
            
        } catch (IOException e) {
            System.err.println("\n✗ Error: " + e.getMessage());
            e.printStackTrace();
            System.exit(1);
        }
    }
    
    /**
     * Create a minimal valid PDF for demonstration.
     */
    private static byte[] createSamplePdfBytes() {
        // Minimal PDF that says "Hello World from PDF!"
        String pdf = "%PDF-1.4\n" +
            "1 0 obj\n<<\n/Type /Catalog\n/Pages 2 0 R\n>>\nendobj\n" +
            "2 0 obj\n<<\n/Type /Pages\n/Kids [3 0 R]\n/Count 1\n>>\nendobj\n" +
            "3 0 obj\n<<\n/Type /Page\n/Parent 2 0 R\n" +
            "/Resources <<\n/Font <<\n/F1 <<\n/Type /Font\n/Subtype /Type1\n" +
            "/BaseFont /Helvetica\n>>\n>>\n>>\n/MediaBox [0 0 612 792]\n" +
            "/Contents 4 0 R\n>>\nendobj\n" +
            "4 0 obj\n<<\n/Length 44\n>>\nstream\n" +
            "BT\n/F1 12 Tf\n100 700 Td\n(Hello World from PDF!) Tj\nET\n" +
            "endstream\nendobj\n" +
            "xref\n0 5\n0000000000 65535 f\n0000000009 00000 n\n" +
            "0000000058 00000 n\n0000000115 00000 n\n0000000317 00000 n\n" +
            "trailer\n<<\n/Size 5\n/Root 1 0 R\n>>\n" +
            "startxref\n410\n%%EOF\n";
        
        return pdf.getBytes();
    }
}

File: clients/python/content_and_search_demo.py
#!/usr/bin/env python3
"""
Document Management System - Content, Renditions, and Search Demo

This demo shows:
1. Creating documents with indexable content
2. Uploading PDF files (simulated)
3. Using transformation features (via UI - not yet exposed in REST)
4. Searching documents with full-text and fielded search
5. Rebuilding the search index
"""

import requests
from docmgmt_client import DocumentManagementClient
import base64


def create_sample_pdf_bytes() -> bytes:
    """
    Create a minimal valid PDF for demonstration.
    In production, you'd use real PDF files.
    """
    # This is a minimal PDF that says "Hello World"
    pdf_content = b"""%PDF-1.4
1 0 obj
<<
/Type /Catalog
/Pages 2 0 R
>>
endobj
2 0 obj
<<
/Type /Pages
/Kids [3 0 R]
/Count 1
>>
endobj
3 0 obj
<<
/Type /Page
/Parent 2 0 R
/Resources <<
/Font <<
/F1 <<
/Type /Font
/Subtype /Type1
/BaseFont /Helvetica
>>
>>
>>
/MediaBox [0 0 612 792]
/Contents 4 0 R
>>
endobj
4 0 obj
<<
/Length 44
>>
stream
BT
/F1 12 Tf
100 700 Td
(Hello World from PDF!) Tj
ET
endstream
endobj
xref
0 5
0000000000 65535 f
0000000009 00000 n
0000000058 00000 n
0000000115 00000 n
0000000317 00000 n
trailer
<<
/Size 5
/Root 1 0 R
>>
startxref
410
%%EOF
"""
    return pdf_content


def main():
    """Demonstrate content management, transformations, and search."""
    print("=== Document Management: Content, Renditions & Search Demo ===\n")
    
    client = DocumentManagementClient()
    
    try:
        # ========================================
        # PART 1: Create Documents with Content
        # ========================================
        print("📄 PART 1: Creating Documents with Indexable Content\n")
        
        # Create a document with text content (indexable)
        print("1. Creating document with text content...")
        doc1 = client.create_document(
            name="Spring Framework Tutorial",
            document_type="ARTICLE",
            description="A comprehensive guide to Spring Boot development",
            keywords="spring boot java framework web",
            tags=["java", "spring", "tutorial", "backend"]
        )
        print(f"   ✓ Created document ID: {doc1['id']}")
        
        # Upload indexable text content
        text_content = """
Spring Framework Tutorial
========================

Spring Boot is a powerful framework for building enterprise Java applications.
It provides features like:
- Dependency Injection
- Auto-configuration
- Embedded servers
- Spring Data for database access
- Spring Security for authentication
        
This makes it ideal for microservices and web applications.
        """.strip()
        
        content1 = client.upload_content_bytes(
            filename="spring_tutorial.txt",
            content_bytes=text_content.encode('utf-8'),
            sys_object_id=doc1['id'],
            store_in_database=True,
            content_type="text/plain"
        )
        print(f"   ✓ Uploaded text content ID: {content1['id']}")
        
        # Create another document
        print("\n2. Creating document with PDF content...")
        doc2 = client.create_document(
            name="Python Best Practices",
            document_type="MANUAL",
            description="Python programming best practices and design patterns",
            keywords="python programming design patterns",
            tags=["python", "best-practices", "programming"]
        )
        print(f"   ✓ Created document ID: {doc2['id']}")
        
        # Upload PDF content (simulated)
        pdf_bytes = create_sample_pdf_bytes()
        content2 = client.upload_content_bytes(
            filename="python_guide.pdf",
            content_bytes=pdf_bytes,
            sys_object_id=doc2['id'],
            store_in_database=True,
            content_type="application/pdf"
        )
        print(f"   ✓ Uploaded PDF content ID: {content2['id']}")
        
        # Create a third document for search testing
        print("\n3. Creating additional document...")
        doc3 = client.create_document(
            name="Database Design Principles",
            document_type="MANUAL",
            description="SQL and NoSQL database design fundamentals",
            keywords="database sql nosql design optimization",
            tags=["database", "sql", "design"]
        )
        
        content3_text = """
Database Design Principles
=========================

Effective database design is crucial for application performance.

Key principles:
1. Normalization - Organize data to reduce redundancy
2. Indexing - Speed up query performance
3. Relationships - Define clear entity relationships
4. Scalability - Plan for growth

Both SQL and NoSQL databases have their place depending on requirements.
        """.strip()
        
        content3 = client.upload_content_bytes(
            filename="database_guide.txt",
            content_bytes=content3_text.encode('utf-8'),
            sys_object_id=doc3['id'],
            store_in_database=True,
            content_type="text/plain"
        )
        print(f"   ✓ Created document ID: {doc3['id']} with text content")
        
        # ========================================
        # PART 2: Transformations & Renditions
        # ========================================
        print("\n\n📝 PART 2: Content Transformations & Renditions\n")
        
        print("NOTE: Content transformation (PDF → text) is currently done via UI.")
        print("Steps to transform PDF content:")
        print("  1. Go to Documents view in the UI")
        print("  2. Select your document")
        print("  3. In the content panel, find your PDF")
        print("  4. Click the ✨ (magic wand) Transform button")
        print("  5. The system will:")
        print("     - Extract text from the PDF using Apache PDFBox")
        print("     - Create a secondary rendition (text/plain)")
        print("     - Mark it as indexable")
        print("     - Add it to the document\n")
        
        print("After transformation, you'll see:")
        contents = client.get_content_by_sysobject(doc2['id'])
        print(f"   Document {doc2['id']} currently has {len(contents)} content item(s)")
        for c in contents:
            print(f"   - {c['name']} ({c.get('contentType', 'unknown')})")
        
        print("\n   After transformation in UI, you'll have:")
        print("   - python_guide.pdf (application/pdf) [Primary, Not Indexable]")
        print("   - python_guide.pdf.plain (text/plain) [Secondary, Indexable]")
        
        # ========================================
        # PART 3: Search Index Management
        # ========================================
        print("\n\n🔍 PART 3: Search Index Management\n")
        
        print("1. Rebuilding search index...")
        result = client.rebuild_search_index()
        print(f"   ✓ {result}")
        
        print("\n2. Checking index statistics...")
        stats = client.get_search_stats()
        print(f"   Documents indexed: {stats.get('documentCount', 0)}")
        print(f"   Max doc ID: {stats.get('maxDoc', 0)}")
        print(f"   Deleted docs: {stats.get('deletedDocs', 0)}")
        
        # ========================================
        # PART 4: Search Operations
        # ========================================
        print("\n\n🔎 PART 4: Searching Documents\n")
        
        # Simple search
        print("1. Simple search for 'spring'...")
        results = client.search("spring", limit=10)
        print(f"   Found {len(results)} result(s):")
        for r in results:
            print(f"   - Doc {r['documentId']}: {r.get('name', 'N/A')} (score: {r['score']:.2f})")
        
        # Search in description
        print("\n2. Search for 'framework'...")
        results = client.search("framework", limit=10)
        print(f"   Found {len(results)} result(s):")
        for r in results:
            print(f"   - Doc {r['documentId']}: {r.get('name', 'N/A')}")
        
        # Search in content (text was indexed)
        print("\n3. Search for 'microservices' (in content)...")
        results = client.search("microservices", limit=10)
        print(f"   Found {len(results)} result(s):")
        for r in results:
            print(f"   - Doc {r['documentId']}: {r.get('name', 'N/A')}")
        
        # Fielded search - search only in tags
        print("\n4. Field-specific search (tags='python')...")
        results = client.search_fields(
            field_queries={"tags": "python"},
            operator="AND",
            limit=10
        )
        print(f"   Found {len(results)} result(s):")
        for r in results:
            print(f"   - Doc {r['documentId']}: {r.get('name', 'N/A')}")
            if r.get('tags'):
                print(f"     Tags: {r['tags']}")
        
        # Multiple field search with AND
        print("\n5. Multi-field search (name='database' AND keywords='sql')...")
        results = client.search_fields(
            field_queries={
                "name": "database",
                "keywords": "sql"
            },
            operator="AND",
            limit=10
        )
        print(f"   Found {len(results)} result(s):")
        for r in results:
            print(f"   - Doc {r['documentId']}: {r.get('name', 'N/A')}")
        
        # Multiple field search with OR
        print("\n6. Multi-field search (name='Spring' OR name='Python' - OR logic)...")
        results = client.search_fields(
            field_queries={
                "name": "Spring Python"  # Search for either term
            },
            operator="OR",
            limit=10
        )
        print(f"   Found {len(results)} result(s):")
        for r in results:
            print(f"   - Doc {r['documentId']}: {r.get('name', 'N/A')}")
        
        # Phrase search
        print("\n7. Phrase search for 'best practices'...")
        results = client.search('"best practices"', limit=10)
        print(f"   Found {len(results)} result(s):")
        for r in results:
            print(f"   - Doc {r['documentId']}: {r.get('name', 'N/A')}")
        
        # Wildcard search
        print("\n8. Wildcard search for 'databa*'...")
        results = client.search("databa*", limit=10)
        print(f"   Found {len(results)} result(s):")
        for r in results:
            print(f"   - Doc {r['documentId']}: {r.get('name', 'N/A')}")
        
        # ========================================
        # PART 5: Working with Renditions
        # ========================================
        print("\n\n📦 PART 5: Inspecting Content Renditions\n")
        
        print(f"1. Content items for document {doc1['id']} (text document):")
        doc1_contents = client.get_content_by_sysobject(doc1['id'])
        for c in doc1_contents:
            print(f"   - {c['name']}")
            print(f"     Type: {c.get('contentType', 'unknown')}")
            print(f"     Storage: {c.get('storageType', 'unknown')}")
            # Note: isPrimary and isIndexable would be in the response if the DTO includes them
        
        print(f"\n2. Content items for document {doc2['id']} (PDF document):")
        doc2_contents = client.get_content_by_sysobject(doc2['id'])
        for c in doc2_contents:
            print(f"   - {c['name']}")
            print(f"     Type: {c.get('contentType', 'unknown')}")
        
        # ========================================
        # Summary
        # ========================================
        print("\n\n" + "="*60)
        print("📊 DEMO SUMMARY")
        print("="*60)
        print(f"✓ Created {3} documents")
        print(f"✓ Uploaded {3} content items")
        print(f"✓ Demonstrated search across multiple fields")
        print(f"✓ Showed fielded search with AND/OR operators")
        print(f"✓ Used phrase and wildcard search")
        print("\n💡 KEY FEATURES:")
        print("   • All document fields are indexed (name, description, keywords, tags)")
        print("   • Text content is automatically indexed")
        print("   • PDF → text transformation via UI (creates indexable renditions)")
        print("   • Primary/secondary rendition support")
        print("   • Full Lucene query syntax support")
        print("   • Field-specific and combined search")
        print("\n📝 NEXT STEPS:")
        print("   1. Transform PDF content via UI to create text renditions")
        print("   2. Rebuild index after transformations")
        print("   3. Search will include PDF text content")
        print("   4. Try advanced Lucene queries: 'spring AND java', 'name:Tutorial', etc.")
        
    except requests.exceptions.RequestException as e:
        print(f"\n✗ Error: {e}")
        if hasattr(e, 'response') and e.response is not None:
            print(f"  Response: {e.response.text}")
        return 1
    
    return 0


if __name__ == "__main__":
    exit(main())
